apis/control/v1alpha1/cassandratask_types.go

```diff
@@ -0,0 +1,188 @@
+	RestartPolicy *corev1.RestartPolicy `json:"restartPolicy,omitempty"`
+
+	// TTLSecondsAfterFinished defines how long the completed job will kept before being cleaned up. If not set,
+	// the task will not be cleaned up by the cass-operator.
+	// +optional
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+	// Specifics if this task can be run concurrently with other active tasks. Valid values are:
+	// - "Allow": allows multiple Tasks to run concurrently on Cassandra cluster
+	// - "Forbid" (default): only a single task is executed at once
+	// The "Allow" property is only valid if all the other active Tasks have "Allow" as well.
+	// +optional
+	ConcurrencyPolicy *batchv1.ConcurrencyPolicy `json:"concurrencyPolicy,omitempty"`
+
+	// TODO Add webhook?
+
+	/*
+		spec:
+			datacenter:
+				name: dc2
+				namespace: cass-operator
+			scheduledTime: "nnnnnn"
+			jobs:
+				- name: rebuild-dc1
+				  command: rebuild
+				  args:
+				    source_datacenter: dc1
+			restartPolicy: Never
+			concurrencyPolicy: Forbid
+			ttlSecondsAfterFinished: 86400
+	*/
+}
+
+type CassandraJob struct {
+	Name string `json:"name"`
+
+	// TODO Should this be typed to match what jobs this version of cass-operator can run?
+	Command string `json:"command"`
+
+	// Should command be typed to restrict allowed values in this version?
+	Arguments map[string]string `json:"args,omitempty"`
+}
+
+// CassandraTaskStatus defines the observed state of CassandraJob
+type CassandraTaskStatus struct {
+
+	// TODO Status and Conditions is almost 1:1 to Kubernetes Job's definitions.
+
+	// The latest available observations of an object's current state. When a Job
+	// fails, one of the conditions will have type "Failed" and status true. When
+	// a Job is suspended, one of the conditions will have type "Suspended" and
+	// status true; when the Job is resumed, the status of this condition will
+	// become false. When a Job is completed, one of the conditions will have
+	// type "Complete" and status true.
+	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=atomic
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	// Represents time when the job controller started processing a job. When a
+	// Job is created in the suspended state, this field is not set until the
+	// first time it is resumed. This field is reset every time a Job is resumed
+	// from suspension. It is represented in RFC3339 form and is in UTC.
+	// +optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// Represents time when the job was completed. It is not guaranteed to
+	// be set in happens-before order across separate operations.
+	// It is represented in RFC3339 form and is in UTC.
+	// The completion time is only set when the job finishes successfully.
+	// +optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
```

Comment:
>The completion time is only set when the job finishes successfully.

Is this correct? If I understand `reconcileEveryPodTask` correctly, it looks like the job should eventually reach a completion state even if it has failed.

Reply:
No, that text is copied over from Job. I don't know yet how to approach this, should it be "CompletionTime + Failed count" to notice that job failed or how would one check that it will no longer be executed (there are cases where we won't retry).